File: 2021/day10/solution.py
from typing import List


CORRUPTED_SCORES = {
    ')': 3,
    ']': 57,
    '}': 1197,
    '>': 25137,
}
INCOMPLETE_SCORES = {
    ')': 1,
    ']': 2,
    '}': 3,
    '>': 4,
}
BRACKETS = {
    '(': ')',
    '[': ']',
    '{': '}',
    '<': '>'
}


def get_inputs():
    with open('input.txt') as f:
        data = []
        lines = f.readlines()
        for l in lines:
            data.append(l.strip())
        return data

def score_autocomplete(s: str) -> int:
    total = 0
    for i in range(len(s)):
        total *= 5
        total += INCOMPLETE_SCORES[s[i]]
    
    return total

def score_line(line: str, count_corrupted: bool) -> int:
    stack = []
    for char in line:
        if char in BRACKETS.keys():
            stack.append(char)
        elif len(stack) > 0:
            val = stack.pop()
            if BRACKETS[val] == char:
                continue
            else:
                if count_corrupted:
                    return CORRUPTED_SCORES[char]
                else:
                    return 0
    
    if not count_corrupted:
        remaining = ""
        while stack:
            val = stack.pop()
            remaining += BRACKETS[val]
        return score_autocomplete(remaining)
    else:
        return 0


def get_scores(lines: List[str], count_corrupted: bool) -> int:
    scores = []

    for l in lines:
        s = score_line(l, count_corrupted)
        if count_corrupted:
            scores.append(s)
        elif s != 0:
            scores.append(s)
    
    scores = sorted(scores)
    if count_corrupted:
        return sum(scores)
    else:
        return scores[len(scores) // 2]

p1 = get_scores(lines=get_inputs(), count_corrupted=True)
print(p1)

p2 = get_scores(lines=get_inputs(), count_corrupted=False)
print(p2)
            
                
                
    

File: 2021/day10/solution.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <stack>
#include <set>
#include <map>
#include <stdexcept>

using namespace std;

const set<char> opens = {'(', '<', '[', '{'};
const set<char> closes = {')', '>', ']', '}'};
map<char,char> pairs = {{')', '('}, {'>', '<'}, {']', '['}, {'}', '{'}};
map<char,int> closeScoreMap = {{'(', 1}, {'[', 2}, {'{', 3}, {'<', 4}};

struct Chunks {
    vector<string> incomplete;
    vector<string> corrupted;
};

Chunks getChunks(const string& filePath) {
    ifstream inputFile(filePath);
    vector<string> lines;
    if (!inputFile) {
        cerr << "Failed to open input file." << endl;
        return {};
    }
    string line;
    while (getline(inputFile, line)) {
        lines.push_back(line);
    }

    Chunks chunks;
    for (const auto& l : lines) {
        stack<char> stack;
        string toCommit = l;
        bool incomplete = true;

        for (const char& c : l) {
            if (opens.count(c) > 0) {
                stack.push(c);
            } else if (stack.empty()){
                incomplete = true;
                toCommit = l;
                break;
            } else {
                auto val = stack.top();
                stack.pop();
                if (pairs[c] == val) {
                    continue;
                } else {
                    incomplete = false;
                    toCommit = l;
                    break;
                }
            }
        }

        if (incomplete) {
            chunks.incomplete.push_back(toCommit);
        } else {
            chunks.corrupted.push_back(toCommit);
        }

    }
    return chunks;
}

int scoreCorruptedChunks(const vector<string>& corrupted) {
    int total = 0;
    for (const auto& corrupt : corrupted) {
        stack<char> stack;
        for (const char& c : corrupt) {
            if (opens.count(c) > 0) {
                stack.push(c);
            } else {
                auto openChar = stack.top();
                stack.pop();
                if (pairs[c] == openChar) {
                    continue;
                } else if (c == ')') {
                    total += 3;
                } else if (c == ']') {
                    total += 57;
                } else if (c == '}') {
                    total += 1197;
                } else if (c == '>') {
                    total += 25137;
                } else {
                    throw runtime_error("found bad character on matching");
                }
                    break;
                }
            }
        }
    return total;
}

long long scoreIncompleteChunks(const vector<string>& incomplete) {
    vector<long long> scores = {};
    for (const auto& inc : incomplete) {
        stack<char> stack;

        for (const char& c : inc) {
            if (opens.count(c) > 0) {
                stack.push(c);
            } else {
                stack.pop();
            }
        }

        long long total = 0;
        while (!stack.empty()) {
            total *= 5;
            auto toScore = stack.top();
            stack.pop();
            total += closeScoreMap[toScore];
        }
        scores.push_back(total);
    }

    sort(scores.begin(), scores.end());
    return scores[scores.size() / 2];
}


int main() {
    auto chunks = getChunks("input.txt");

    cout << "Part 1: " + to_string(scoreCorruptedChunks(chunks.corrupted)) << endl;
    cout << "Part 2: " + to_string(scoreIncompleteChunks(chunks.incomplete)) << endl;
}